crates/curp/src/server/raw_curp/log.rs

```diff
@@ -134,36 +179,39 @@ impl<C: Command> LogEntryVecDeque<C> {
 
     /// restore log entries from Vec
     fn restore(&mut self, entries: Vec<LogEntry<C>>) {
-        let mut batch_index = VecDeque::with_capacity(entries.capacity());
-        batch_index.push_back(0);
-        for entry in &entries {
-            #[allow(clippy::expect_used)]
-            let entry_size =
-                serialized_size(entry).expect("log entry {entry:?} cannot be serialized");
-            if let Some(cur_size) = batch_index.back() {
-                batch_index.push_back(cur_size.overflow_add(entry_size));
-            }
-        }
+        self.batch_index = VecDeque::with_capacity(entries.capacity());
+        self.entries = VecDeque::with_capacity(entries.capacity());
+        self.entry_size = VecDeque::with_capacity(entries.capacity());
 
-        self.entries = entries.into_iter().map(Arc::new).collect();
-        self.batch_index = batch_index;
+        self.last_batch_size = 0;
+        self.first_entry_at_last_batch = 0;
+
+        for entry in entries {
+            let _unuse = self.push_back(Arc::from(entry));
+        }
     }
 
     /// clear whole log entries
     fn clear(&mut self) {
         self.entries.clear();
+        self.entry_size.clear();
         self.batch_index.clear();
-        self.batch_index.push_back(0);
+        self.last_batch_size = 0;
+        self.first_entry_at_last_batch = 0;
     }
 
     /// Get the range [left, right) of the log entry, whose size should be equal or smaller than `batch_limit`
+    #[allow(clippy::range_plus_one)]
     fn get_range_by_batch(&self, left: usize) -> Range<usize> {
-        #[allow(clippy::indexing_slicing)]
-        let target = self.batch_index[left].overflow_add(self.batch_limit);
-        // remove the fake index 0 in `batch_index`
-        match self.batch_index.binary_search(&target) {
-            Ok(right) => left..right,
-            Err(right) => left..right - 1,
+        #![allow(clippy::indexing_slicing)]
+        if left >= self.batch_index.len() {
+            left..self.entries.len()
+        } else if self.entry_size[left] > self.batch_limit {
+            left..left + 1
+        } else if self.batch_index[left] == 0 {
+            left..self.entries.len()
+        } else {
+            left..left + self.batch_index[left]
         }
```

Comment:
There's too much if-else statements, please consider refactor this and merge the conditions.

Reply:
Done